mountpoint-s3/src/main.rs

```diff
@@ -600,8 +609,12 @@ fn parse_bucket_name(bucket_name: &str) -> anyhow::Result<String> {
     // Actual bucket names must start/end with a letter, but bucket aliases can end with numbers
     // (-s3), so let's just naively check for invalid characters.
     let bucket_regex = Regex::new(r"^[0-9a-zA-Z\-\._]+$").unwrap();
-    if !bucket_regex.is_match(bucket_name) {
-        return Err(anyhow!("bucket names can only contain letters, numbers, . and -"));
+    // ARN can have `:` and `/` but the prefix is always `arn:`
+    let arn_regex = Regex::new(r#"^arn:aws(-[a-z]+)*:s3(.*):([0-9]+):(.*?)(:(.*?))?$"#).unwrap();
```

Comment:
RegEx can be difficult to understand. Since we care mostly about steering the user in the right direction but it doesn't need to be perfect, I'd recommend this one:

```suggestion
    // Simple ARN check. Doesn't have to be perfect.
    let arn_regex = Regex::new("^\
        arn:\
        aws(-[a-z]+)*:\
        s3(-[a-z]+)*:\
        ([a-z]+(-[a-z0-9]+)+)?:\
        ([0-9]+):\
        .+\
        $").unwrap();
```

I've broken the RegEx onto multiple lines with the `\`, added (optional) region, and simplified to just `.+` after the account ID - I think its clearer to reader now which bit of the ARN is which.

Reply:
I would do something dumber: just accept anything that starts with "arn:" as an ARN. This parser is only here to help detect simple mistakes like providing an "s3://" URL or swapping the directory and bucket name arguments. There's no need to be precise, and it's probably a little more future-proof to not over-specify here.

Reply:
Sounds good to me. I suspect this logic better belongs to the endpoint resolver anyway, if it is checked.